### PR TITLE
Replicate concurrency helper from v2 for v1 limits

### PR DESCRIPTION
**Pull request by abrookins**
This PR uses an alternative approach to support tag-based concurrency limits with client-side task run orchestration. We had been using global limits for tag-based concurrency, but the lack of support for tracking holders of limits made that messy.

Instead of using global limits, this PR introduces new `concurrency` helpers (sync and async) that mirror the `concurrency` helpers for global limits. The difference is that these use existing task run concurrency limits. To make this work, we've extended the task run concurrency HTTP endpoints to support the same batch increment and decrement functionality that global limits support.

Trying all these pieces together, this PR also swaps the new `concurrency` helpers into the task engine to handle tag-based concurrency.

### Checklist

- [x] This pull request references any related issue by including "closes `<link to issue>`"
  - If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect/issues/new/choose) first.
- [x] If this pull request adds new functionality, it includes unit tests that cover the changes
- [x] If this pull request removes docs files, it includes redirect settings in `mint.json`.
- [x] If this pull request adds functions or classes, it includes helpful docstrings.

### CODE REVIEW

File: src/prefect/concurrency/v1/services.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+                        mode=mode,
+                        create_if_missing=create_if_missing,
+                        holder=holder,
+                    )
+                except Exception as exc:
+                    if (
+                        isinstance(exc, httpx.HTTPStatusError)
+                        and exc.response.status_code == status.HTTP_423_LOCKED
+                    ):
+                        retry_after = float(exc.response.headers["Retry-After"])
```
Comment (chrisguidry) [inline src/prefect/concurrency/v1/services.py:85]: One small note, there's a case where we actually need to just stop the task entirely, so in that case you'll get a 423 _without a Retry-After_.  That happens when the limit exists and it actually set to `0`.  I think that's a way people can disable tasks without redeploying, kinda like a big red panic button.

In that case we probably need to raise something that will tell the user why?

File: tests/server/orchestration/api/test_concurrency_limits.py [hunk 2]
```diff
@@ -270,3 +271,79 @@ async def test_setting_tag_to_zero_concurrency(
                 read_response.json()
             )
             assert concurrency_limit.active_slots == []
+
+    async def test_acquiring_returns_limits(
```
Comment (chrisguidry) [inline tests/server/orchestration/api/test_concurrency_limits.py:275]: I love that you did this as a strict addition of new tests :bow:

File: src/prefect/concurrency/v1/services.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+                    ):
+                        retry_after = exc.response.headers.get("Retry-After")
+                        if retry_after:
+                            retry_after = float(retry_after)
+                            await asyncio.sleep(retry_after)
+                        else:
+                            # We received a 423 but no Retry-After header. This
+                            # should indicate that the server told us to abort
+                            # because the concurrency limit is set to 0, i.e.
+                            # effectively disabled.
```
Comment (chrisguidry) [inline src/prefect/concurrency/v1/services.py:87]: :fire:

File: src/prefect/concurrency/v1/services.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+                else:
+                    return response
+
+    def send(self, item: Tuple[UUID, Optional[float]]) -> concurrent.futures.Future:
+        with self._lock:
+            if self._stopped:
+                raise RuntimeError("Cannot put items in a stopped service instance.")
+
+            logger.debug("Service %r enqueuing item %r", self, item)
+            future: concurrent.futures.Future = concurrent.futures.Future()
```
Comment (chrisguidry) [inline src/prefect/concurrency/v1/services.py:111]:
```suggestion
            future = concurrent.futures.Future()
```

The type of the variable is implied by the type of the rvalue here

**Review by chrisguidry — APPROVED** [review]
Looks great!  The test suite is great, thank you!

File: src/prefect/concurrency/v1/asyncio.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+        timeout_seconds: The number of seconds to wait for the slots to be acquired before
+            raising a `TimeoutError`. A timeout of `None` will wait indefinitely.
+
+    Raises:
+        TimeoutError: If the slots are not acquired within the given timeout.
+
+    Example:
+    A simple example of using the async `concurrency` context manager:
+    ```python
+    from prefect.concurrency.asyncio import concurrency
```
Comment (desertaxle) [inline src/prefect/concurrency/v1/asyncio.py:57]:
```suggestion
    from prefect.concurrency.v1.asyncio import concurrency
```

**Review by desertaxle — APPROVED** [review]

**Comment by abrookins** [comment]
I can't get tests to pass, but the problem [appears to be in main](https://github.com/PrefectHQ/prefect/actions/runs/10509231964/job/29114933467), too, so I'm going to merge.

**Pull request merged**